nexus/src/app/mod.rs

```diff
@@ -153,6 +153,24 @@ impl Nexus {
         let db_datastore = Arc::new(db::DataStore::new(Arc::clone(&pool)));
         db_datastore.register_producers(&producer_registry);
 
+        // Keep looping until we find that the schema matches our expectation.
+        const EXPECTED_VERSION: SemverVersion = SemverVersion::new(1, 0, 0);
+        loop {
+            match db_datastore.database_schema_version().await {
+                Ok(version) => {
+                    if version == nexus_db_model::schema::SCHEMA_VERSION {
+                        break;
+                    }
+                    let observed = version.0;
+                    warn!(log, "Incompatible database schema: Saw {observed}, expected {EXPECTED_VERSION}");
+                }
+                Err(e) => {
+                    warn!(log, "Cannot read database schema version: {e}");
+                }
+            }
+            tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+        }
```

Comment:
I'm open to putting this in a different spot, but it seems necessary to do it before we start the saga executor / start populating / start background tasks.